ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/SubordinationManagerXAResourceOrphanFilter.java

```diff
@@ -0,0 +1,93 @@
+ * See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ * This copyrighted material is made available to anyone wishing to use,
+ * modify, copy, or redistribute it subject to the terms and conditions
+ * of the GNU Lesser General Public License, v. 2.1.
+ * This program is distributed in the hope that it will be useful, but WITHOUT A
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License,
+ * v.2.1 along with this distribution; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA  02110-1301, USA.
+ *
+ * (C) 2016,
+ * @author JBoss, by Red Hat.
+ */
+package com.arjuna.ats.internal.jta.recovery.arjunacore;
+
+import com.arjuna.ats.arjuna.coordinator.TxControl;
+import com.arjuna.ats.arjuna.recovery.RecoveryManager;
+import com.arjuna.ats.arjuna.recovery.RecoveryModule;
+import com.arjuna.ats.internal.jta.transaction.arjunacore.jca.SubordinateTransaction;
+import com.arjuna.ats.internal.jta.transaction.arjunacore.jca.SubordinationManager;
+import com.arjuna.ats.jta.recovery.XAResourceOrphanFilter;
+import com.arjuna.ats.jta.xa.XATxConverter;
+import com.arjuna.ats.jta.xa.XidImple;
+
+import javax.transaction.xa.XAException;
+import javax.transaction.xa.Xid;
+import java.util.Vector;
+
+/**
+ * An XAResourceOrphanFilter which vetos rollback for xids which have an in-flight subordinate transaction.
+ * <p>
+ * The SubordinateAtomicActionRecoveryModule must be loaded in in a position prior to the XARecoveryModule for this to work
+ */
+public class SubordinationManagerXAResourceOrphanFilter implements XAResourceOrphanFilter {
+    private boolean validated;
+
+    @Override
+    public Vote checkXid(Xid xid) {
+        if (xid.getFormatId() != XATxConverter.FORMAT_ID) {
+            // we only care about Xids created by the JTA
+            return Vote.ABSTAIN;
+        }
+
+        if (!TxControl.getXANodeName().equals(XATxConverter.getSubordinateNodeName(new XidImple(xid).getXID()))) {
+            // It either doesn't have a subordinate node name or isn't for this server
+            return Vote.ABSTAIN;
+        }
+
+        if (!validatePosition()) {
+            return Vote.LEAVE_ALONE;
+        }
+
+        XidImple theXid = new XidImple(xid);
+        SubordinateTransaction importedTransaction = null;
+        try {
+            importedTransaction = SubordinationManager.getTransactionImporter().getImportedTransaction(theXid);
+        } catch (XAException e) {
+            return Vote.LEAVE_ALONE;
+        }
+        if (importedTransaction != null) {
+            return Vote.LEAVE_ALONE;
+        } else {
+            return Vote.ROLLBACK;
+        }
+    }
+
+    private boolean validatePosition() {
```

Comment:
This complicated (and probably non performant) routine is essentially just doing the following:
```java
return modules.contains(SAARM) && modules.contains(XARM))
```
Can we either refactor the code or simplify.

Reply:
It's checking the order too (hence the name) - I can add more comments